### ISSUE TITLE
Train a segement modle

### ISSUE THREAD

**Issue by skylning** [lightly-ai/lightly-train#142]
How to use dinov2 method to train a segemnt model by lightly-train

**Comment by liopeer** [comment]
Hi @skylning and thank you for your question!

We currently don't have a tutorial for semantic segmentation, however we do have one for [depth estimation with a UNet](https://docs.lightly.ai/train/stable/tutorials/depth_estimation/index.html), which is a related task since it also gives pixel-wise predictions. During fine-tuning you would of course need to change the loss function to a cross-entropy loss and you would have to adapt the datasets. Are you interested in using a particular architecture, e.g. something Transformer-based? And are you planning to use open datasets for pretraining and finetuning? If yes, we could create a tutorial directly for your use case, since we were planning to add a tutorial for semantic segmentation anyway.

**Comment by skylning** [comment]
thank you ,I am looking forward to the segmentation example

**Comment by kaushikCanada** [comment]
i am trying to use this for land cover segmentation

**Comment by liopeer** [comment]
Small update: In [0.10.0](https://github.com/lightly-ai/lightly-train/releases/tag/v0.10.0) we added support for training the EoMT model for semantic segmentation.

The instructions you can find here in the docs: https://docs.lightly.ai/train/stable/semantic_segmentation.html

Let me know if this already covers your use case. If yes, I would go ahead and close this issue.

**Comment by kaushikCanada** [comment]
Thank you so much for this. I am trying to compare some models for land cover semantic segentation. 
model1 - resnet 18 unet without pretraining
model2 - resnet 18 unet with dinov2 distilaltion
model3 - EOMT with Dinov2 
model4 - EOMT with Dinov3

. My question is is this kind of good enuf ? let me know if i can get more out f lightlytrain to compare (like should i put DINO also?). I have 48gb GPU. and second is Is there a way to do the embedding thing since i am doing segmentation? I am jsut guessing DInov3 is going to bethe best?

**Comment by liopeer** [comment]
I think you will see very good results with EoMT and DINOv3 – our internal benchmarks are so far really promising! If your imagery is satellite-based you might also consider using the DINOv3 checkpoints that are trained on [SAT493m](https://ai.meta.com/resources/models-and-libraries/dinov3-downloads/).

I expect quite a performance gap between the ResNet-based UNets and EoMT, so if you want something in between to compare to it might be worth looking at DeepLabv3+ (for something convolutional) or SegFormer (for something Transformer-based). I think with a 48Gb GPU you should be able to train everything that I mentioned, however pretraining with distillation might take quite some time on a single GPU.

**Comment by liopeer** [comment]
If you successfully use the SAT493m checkpoints, let me know how well they worked. I am personally interested in them as well!

**Comment by kaushikCanada** [comment]
I have applied for DinoV3 access but not received any email yet.

**Comment by liopeer** [comment]
It took a few hours for me as well. @kaushikCanada

**Comment by kaushikCanada** [comment]
Thank you Lionel. Also I want to train a DinoV2 from scratch and then put a
linear segmentation head on it. Also, I have both RGB and multispectral
data. How can I do these things if you can give me some advice?

**Comment by liopeer** [comment]
Unfortunately, that particular use case is not supported at the moment, since you currently can't modify the number of input channels in LightlyTrain. We plan to support this in the near future, but it might still take a few weeks (feel free to open a separate issue for this).

However, I can recommend to you the work [*Panopticon*](https://github.com/Panopticon-FM/panopticon) from the recent CVPR2025 which should solve exactly your problem. It's also included in TorchGeo: https://torchgeo.readthedocs.io/en/stable/api/models.html#panopticon

**Comment by kaushikCanada** [comment]
Hello Lionel. Thank you for your suggestion. Yes i saw the pantopticon and
was wondering if I should do that. Seems like a big model :). I love
torchgeo btw and have been attempting terratorch recently. For now i have
rgb images from drone for the whole study area. Can i just export them as
224x224 and simply run the lightly train on that? my labels are land cover
classes.

**Comment by liopeer** [comment]
No need to export at 224 resolution. LightlyTrain will crop them automatically to the correct resolution (which is 518x518 for EoMT and 224x224 for pretraining, i.e. DINOv2 distillation).

**Comment by kaushikCanada** [comment]
Woww marvellous. So i can keep my original tile at 6000x6000 and lightly
will crop things out? That would be great. Also i have previously created
overlapped patches with like say 50% overlap with unet. Do i need to have
anything like that here?

A question regarding the label classes, my claases are in integer masks
with values like 1 4 10 15 etc. Will i put these values in the dictionary
as is? And in some images i have nodata values which is 0 in the input and
0 in the output. So i should put 0 in the ignore index?

**Comment by liopeer** [comment]
> Woww marvellous. So i can keep my original tile at 6000x6000 and lightly
will crop things out?

Theoretically yes, however if they are this large the decoding might take a bit of time and you could end up CPU bottle-necked. With a single GPU I don't think that's a serious risk, but maybe something to be aware of

> Also i have previously created
overlapped patches with like say 50% overlap with unet. Do i need to have
anything like that here?

No need during training. We will randomly crop to 518x518 to ensure that everything will eventually be covered. If you want to eventually run inference on 6000x6000 images however you might have to run it on sub-patches as well and then stitch the results together.

> A question regarding the label classes, my claases are in integer masks
with values like 1 4 10 15 etc. Will i put these values in the dictionary
as is?

Yes, you can insert them like this. Internally we will remap them to indices in `{0, .., N}`, but you the result of the inference will again be in your indices.

> And in some images i have nodata values which is 0 in the input and
0 in the output. So i should put 0 in the ignore index?

Depends: If you put 0 in the `ignore_index`, that means that this class will not contribute to the loss during training. It's often used for "background" or "everything else". In your case, if 0 means "no land cover", it might be worth keeping. However, I am not an expert on this.

**Comment by kaushikCanada** [comment]
Wow thank you so much this is really helpful. One thigh that i just saw is
in lightly train the dinov2/v3 backbone url is like
https://dinov3.llamameta.net/dinov3_vits16. however i applied to the
official link and have not received reply yet.  So i went to huggingface
for facebook/dinov3-vitl16-pretrain-sat493m and sent a request on that
portal and i promptly received approval. is it possible to use the
huggingface model somehow or we need the official url?

**Comment by liopeer** [comment]
The Huggingface checkpoints are in the SafeTensors format and therefore not standard torch checkpoints. I fear that it might not work out of the box. But if you would like support for that I recommend that you open a separate issue and then we can maybe take care of supporting this in a few weeks.
